### PR TITLE
Refactor EventTimer + Changes to commercial metrics

### PR DESCRIPTION
**Pull request by Jakeii**
## What does this change?
Refactor event-timer.ts with the following changes

- Remove google analytics reporting as we don't use it any longer
- Change the way that timers are stored (use a map rather than an array/object combo)
- Remove the first timer, and measure specific slots instead (see below)
- Introduce some measures/deltas to replace some of the timers/marks

### First timer removal
Using the `first` for each event has led to some inconsistencies in the data, for roughly 21% of data points the timers are recorded in the wrong order, we believe could be caused by the first event of each type coming from different slots in some circumstances.

In most cases the first slot is simply top-above-nav anyway, and using pageviews where that isn't the case could be adding some undesirable noise/variables to the data.

We also have very little insight into what goes on on ad slots further down the page, we we will add reporting for inline1 (often a video ad) and inline2 (floated right on desktop).

### Deltas & new measurements
Most of the timers we don't actually even use for analysis, only `adOnPage` generally. Changing most of them to deltas will give us more useful information more easily about specific pieces of code that we're interested in e.g. how long does prebid, defineSlot (which incorporates IAS), loadSlot (fetching the ad from GAM) take.

Hopefully deltas will be superior timestamps as they won't be subjected to variance from before the timer we're interested in has even started.

What we're doing in diagram form,  we are keep commercialStart and adOnPage (green) as timestamps, throwing away the rest (red) and adding delta's (in blue)
new-timers

new-timers2

### Summary of deltas
```
commercialBoot			-> time for commercial modules to load
googletagInit			-> Time for googletag to load
${slot}-defineSlot		-> Defining the slot (`window.googletag.defineSlot`) and running IAS
${slot}-prepareSlot		-> mostly Prebid.js, but includes our own code around it
${slot}-prebid			-> the time for prebid auction to happen
${slot}-fetchAd			-> the time for the ad to be fetched from GAM
${slot}-adRender		-> the time for the ad to render, from when it's in view to the ad appearing on the page (should be loadSlot + adRender)
```

### Summary of timers
```
commercialStart         -> Pretty much the earliest point in the commercial code
commercialModulesLoaded -> All async modules in commercial.standalone.ts loaded
${slot}-adSlotReady		-> Ad slot defined and ready to render
${slot}-adRenderStart 	-> Ad (scrolled) in view and loading
${slot}-adOnPage    	-> The ad appeared on the page

```

### CODE REVIEW

**Review by chrislomaxjones — COMMENTED** [review]
Thanks for the very helpful diagram!

File: src/core/event-timer.ts [hunk 1]
```diff
@@ -183,71 +197,67 @@ class EventTimer {
...
-			typeof window.performance.mark === 'function'
-		) {
-			const mark = window.performance.mark(longName);
 			// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- browser support is patchy
-			if (typeof mark?.startTime === 'number') {
-				this._events.push(new Event(name, mark));
-			}
+			typeof mark?.startTime === 'number' &&
+			// we only want to save the marks that are related to certain slots or the page
+			shouldSaveMark(name)
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts:223]: The idea was to have all the marks and measures appear in the inspector's performance tab for local debugging but only send certain one's to the datalake.
There's probably still a better way to do this without joining and pulling apart the names though!

File: src/core/event-timer.ts [hunk 2]
```diff
@@ -49,20 +21,86 @@ interface EventTimerProperties {
...
+type PageMeasure = (typeof pageMeasures)[number];
+
+// all marks, including the measure start and end marks
+const allSlotMarks = [
+	...slotMarks,
+	...slotMeasures.map((measure) => `${measure}Start`),
+	...slotMeasures.map((measure) => `${measure}End`),
+] as const;
+
+enum ExternalEvents {
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts]: yeah makes sense!

File: src/core/event-timer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrislomaxjones) [reply src/core/event-timer.ts:221]: Got it, so we might want to trigger some events that we don't want to ship to the data warehouse?

> There's probably still a better way to do this without joining and pulling apart the names though!

We can always refactor later 🙂

**Review by chrislomaxjones — APPROVED** [review]
Nice one!

**Review by arelra — COMMENTED** [review]
Nice work and thanks for the thorough explanation! Just some comments on naming conventions that we can discuss.

**Comment by domlander** [comment]
Could you update the handbook when this is live? 

https://docs.google.com/document/d/10Mxxta6ypxAduVXJFp19uRoWtALQmAubdKtBwq7AqZ0/edit#heading=h.w0c6dt3g4krt

File: src/core/event-timer.ts [hunk 3]
```diff
@@ -49,20 +21,88 @@ interface EventTimerProperties {
 	labsUrl?: string;
 }
 
-class EventTimer {
-	private _events: Event[];
-	private static _externallyDefinedEventNames = [
-		'cmp-init',
-		'cmp-ui-displayed',
-		'cmp-got-consent',
+// Events will be logged using the performance API for all slots, but only these slots will be tracked as commercial metrics
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts:24]: yes, I'll clarify this comment, only the slots listed here will be sent to the data lake

File: src/core/event-timer.ts [hunk 4]
```diff
@@ -183,71 +195,67 @@ class EventTimer {
...
+	 * For slot events also ensures each TYPE of event event is only logged once per slot
+	 *
+	 * TODO more strict typing for eventName and origin
 	 *
-	 * @param {string} eventName - The short name applied to the mark
-	 * @param {origin} [origin=page] - Either 'page' (default) or the name of the slot
+	 * @param eventName The short name applied to the mark
+	 * @param origin - Either 'page' (default) or the name of the slot
 	 */
 	trigger(eventName: string, origin = 'page'): void {
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts:205]: yeah, makes sense

File: src/core/event-timer.ts [hunk 5]
```diff
@@ -49,20 +21,88 @@ interface EventTimerProperties {
...
+// marks that we want to save as commercial metrics
+const slotMarks = ['slotReady', 'adRenderStart', 'adOnPage'] as const;
+
+type SlotMark = (typeof slotMarks)[number];
+
+// measures that we want to save as commercial metrics
+const slotMeasures = [
+	'adRender',
+	'defineSlot',
+	'prepareSlot',
```
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts:38]: I changed my mind about the name at some point, will update the diagram

File: src/core/event-timer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jakeii) [reply src/core/event-timer.ts]: Exactly, I'll at least update the comment to clarify a bit better :)

**Review by arelra — APPROVED** [review]
Great work! 🎉

**Review by chrislomaxjones — APPROVED** [review]
Looking forward to seeing these timers roll out!

**Pull request merged**